Tests/Update-GitBranch.Tests.ps1

```diff
@@ -0,0 +1,291 @@
+        $Global:Error | Should -Match $Pattern
+    }
+}
+
+function ThenHeadIsLastCommit
+{
+    param(
+        $BranchName = 'master'
+    )
+
+    $repo = Get-GitRepository -RepoRoot $clientDirectory
+    try
+    {
+        It ('should not create new commit') {
+            $repo.Branches[$BranchName].Tip.Sha | Should -Be $lastCommit.Sha
+        }
+    }
+    finally
+    {
+        $repo.Dispose()
+    }
+}
+
+function ThenHeadIsNewCommit
+{
+    $repo = Get-GitRepository -RepoRoot $clientDirectory
+    try
+    {
+        It ('should create new commit') {
+            $head = $repo.Branches['master'].Tip
+            $head.Sha | Should -Not -Be $lastCommit.Sha
+            $head.Parents | Where-Object { $_.Sha -eq $lastCommit.Sha } | Should -Not -BeNullOrEmpty
+        }
+    }
+    finally
+    {
+        $repo.Dispose()
+    }
+}
+
+function ThenStatusIs
+{
+    param(
+        [LibGit2Sharp.MergeStatus]
+        $ExpectedStatus
+    )
+    
+    It ('should result in "{0}" merge' -f $ExpectedStatus) {
+        $result.Status | Should -Be $ExpectedStatus
+    }
+}
+
+function ThenUpdateFailed
+{
+    It ('should fail the update') {
+        $script:result | Should -BeNullOrEmpty
+        $Global:Error | Should -Not -BeNullOrEmpty
+    }
+}
+
+function WhenUpdated
+{
+    [CmdletBinding()]
+    param(
+        $RepoRoot,
+        $AndMergeStrategyIs
+    )
+
+    $mergeStrategyArg = @{}
+    if( $AndMergeStrategyIs )
+    {
+        $mergeStrategyArg['MergeStrategy'] = $AndMergeStrategyIs
+    }
+
+    $script:result = Update-GitBranch -RepoRoot $RepoRoot @mergeStrategyArg
+}
+
+Describe 'Update-GitBranch.when no new commits on the server' {
+    Init
+    GivenNewCommitIn $clientDirectory
+    WhenUpdated -RepoRoot $clientDirectory
+    ThenStatusIs 'UpToDate'
+    ThenHeadIsLastCommit
+}
+
+Describe 'Update-GitBranch.when no new commits local and no new commits on server' {
+    Init
+    WhenUpdated -RepoRoot $clientDirectory
+    ThenStatusIs 'UpToDate'
+    ThenHeadIsLastCommit
+}
+
+Describe 'Update-GitBranch.when no new commits local and new commits on server' {
+    Init
+    GivenNewCommitIn $serverWorkingDirectory -AndPushed
+    WhenUpdated -RepoRoot $clientDirectory
+    ThenStatusIs 'FastForward'
+    ThenHeadIsLastCommit
+}
+
+Describe 'Update-GitBranch.when no new commits local and new commits on server' {
+    Init
+    GivenNewCommitIn $serverWorkingDirectory -AndPushed
+    WhenUpdated -RepoRoot $clientDirectory -AndMergeStrategyIs 'Merge'
+    ThenStatusIs 'NonFastForward'
+    ThenHeadIsNewCommit
+}
+
+Describe 'Update-GitBranch.when new commits local and new commits on server' {
+    Init
+    GivenNewCommitIn $serverWorkingDirectory -AndPushed
+    GivenNewCommitIn $clientDirectory
+    WhenUpdated -RepoRoot $clientDirectory
+    ThenStatusIs 'NonFastForward'
+    ThenHeadIsNewCommit
+}
+
+Describe 'Update-GitBranch.when new commits local and new commits on server and merge must be fast-forwarded' {

```

Comment:
Do we need to add a test where the merge strategy is `Merge` but the server/client commits contain a conflict?

Reply:
Test added. Ready for re-review. Size of changes has gone down, though.